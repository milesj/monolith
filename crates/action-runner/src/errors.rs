use moon_error::MoonError;
use moon_project::ProjectError;
use moon_workspace::WorkspaceError;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum ActionRunnerError {
    #[error("{0}")]
    Failure(String),

    #[error(transparent)]
    DepGraph(#[from] DepGraphError),

    #[error(transparent)]
    Moon(#[from] MoonError),

    #[error(transparent)]
    Project(#[from] ProjectError),

    #[error(transparent)]
    Workspace(#[from] WorkspaceError),
}

#[derive(Error, Debug)]
pub enum DepGraphError {
    #[error("A dependency cycle has been detected for <file>{0}</file>.")]
    CycleDetected(String),

    #[error("Unknown node {0} found in dependency graph. How did this get here?")]
    UnknownNode(usize),

    #[error(transparent)]
    Moon(#[from] MoonError),

    #[error(transparent)]
    Project(#[from] ProjectError),
}
