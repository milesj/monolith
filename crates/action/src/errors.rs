use moon_error::MoonError;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum ActionError {
    #[error(transparent)]
    Moon(#[from] MoonError),
}
