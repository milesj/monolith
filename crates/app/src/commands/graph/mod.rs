pub mod action;
pub mod dep;
mod dto;
pub mod project;
pub mod task;
mod utils;
