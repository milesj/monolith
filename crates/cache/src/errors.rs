use thiserror::Error;

#[derive(Error, Debug)]
pub enum CacheError {
    #[error(transparent)]
    IO(#[from] std::io::Error),

    #[error(transparent)]
    JSON(#[from] serde_json::Error),
}
