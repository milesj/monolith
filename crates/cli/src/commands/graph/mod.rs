pub mod action;
pub mod dep;
mod dto;
pub mod project;
mod utils;
