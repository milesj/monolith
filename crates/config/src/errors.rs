use figment::error::Kind;
use figment::Error as FigmentError;
use std::borrow::Cow;
use validator::{ValidationError, ValidationErrors};

pub fn create_validation_error(code: &'static str, path: &str, message: String) -> ValidationError {
    let mut error = ValidationError::new(code);
    error.message = Some(Cow::from(message));
    // Is there a better way to do this?
    error.add_param(Cow::from("path"), &path.to_owned());
    error
}

pub fn map_figment_error_to_validation_errors(figment_error: &FigmentError) -> ValidationErrors {
    let path = figment_error.path.join(".");

    let valid_error = match &figment_error.kind {
        // Fields
        Kind::DuplicateField(field) => create_validation_error(
            "duplicate_field",
            path.as_str(),
            format!("Duplicate field `{}`.", field),
        ),
        Kind::MissingField(field) => create_validation_error(
            "missing_field",
            path.as_str(),
            format!("Missing field `{}`.", field),
        ),
        Kind::UnknownField(field, _) => create_validation_error(
            "unknown_field",
            path.as_str(),
            format!("Unknown field `{}`.", field),
        ),
        Kind::UnknownVariant(field, _) => create_validation_error(
            "unknown_field_variant",
            path.as_str(),
            format!("Unknown option `{}`.", field),
        ),

        // Values
        Kind::InvalidType(a, e) => create_validation_error(
            "invalid_type",
            path.as_str(),
            format!("Expected {} type, received {}.", e, a),
        ),
        Kind::InvalidLength(a, e) => create_validation_error(
            "invalid_length",
            path.as_str(),
            format!("Expected length of {}, received {}.", e, a),
        ),
        Kind::InvalidValue(a, e) => create_validation_error(
            "invalid_value",
            path.as_str(),
            format!("Expected {} value, received {}.", e, a),
        ),
        Kind::ISizeOutOfRange(range) => create_validation_error(
            "out_of_range",
            path.as_str(),
            format!("Integer out of range, received {}.", range),
        ),
        Kind::USizeOutOfRange(range) => create_validation_error(
            "out_of_range",
            path.as_str(),
            format!("Unsigned integer out of range, received {}.", range),
        ),

        // Other
        Kind::Message(message) => {
            create_validation_error("message", path.as_str(), String::from(message))
        }
        Kind::Unsupported(a) => create_validation_error(
            "unsupported",
            path.as_str(),
            format!("Unsupported type/value `{}`.", a),
        ),
        Kind::UnsupportedKey(key, _) => create_validation_error(
            "unsupported_key",
            path.as_str(),
            format!("Unsupported key `{}`.", key),
        ),
    };

    let mut errors = ValidationErrors::new();

    // We basically need a string literal here, but the path is dynamically provided...
    // https://stackoverflow.com/a/52367953
    errors.add(
        Box::leak(figment_error.path.join(".").into_boxed_str()),
        valid_error,
    );

    errors
}

#[cfg(test)]
pub mod tests {
    use figment::error::Kind;
    use figment::Error;
    use serde_yaml::{to_value, Value};
    use validator::{ValidationError, ValidationErrors, ValidationErrorsKind};

    fn format_yaml_value(value: Value) -> String {
        match value {
            Value::Null => String::from("null"),
            Value::Bool(_) => String::from("boolean"),
            Value::Number(_) => String::from("number"),
            Value::String(msg) => msg,
            Value::Sequence(array) => format!("array of {:?}", array),
            Value::Mapping(object) => format!("object of {:?}", object),
        }
    }

    fn format_validation_error(error: &ValidationError) -> String {
        let mut message = "".to_owned();

        if let Some(path) = error.params.get("path") {
            let value = format_yaml_value(to_value(&path).unwrap());

            if !value.is_empty() {
                let msg = format!("Invalid field `{}`. ", value);
                message.push_str(msg.as_str());
            }
        }

        if error.message.is_some() {
            let msg = format!("{}", error.message.as_ref().unwrap());
            message.push_str(msg.as_str());
        } else {
            let msg = format!("Unknown failure [{}].", error.code);
            message.push_str(msg.as_str());
        }

        message
    }

    fn extract_first_error(errors: &ValidationErrors) -> String {
        for val in errors.errors().values() {
            match val {
                ValidationErrorsKind::Struct(obj) => {
                    let result = extract_first_error(obj);

                    if !result.is_empty() {
                        return result;
                    }
                }
                ValidationErrorsKind::List(list) => {
                    if !list.is_empty() {
                        let item = extract_first_error(list.values().next().unwrap());

                        if !item.is_empty() {
                            return item;
                        }
                    }
                }
                ValidationErrorsKind::Field(field) => {
                    if !field.is_empty() {
                        return format_validation_error(&field[0]);
                    }
                }
            }
        }

        String::from("")
    }

    pub fn handled_jailed_error(errors: &ValidationErrors) -> Error {
        Error::from(Kind::Message(extract_first_error(errors)))
    }
}
