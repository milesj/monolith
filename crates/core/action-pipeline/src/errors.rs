use miette::Diagnostic;
use moon_dep_graph::DepGraphError;
use moon_error::MoonError;
use moon_project::ProjectError;
use moon_runner::RunnerError;
use moon_target::TargetError;
use moon_tool::ToolError;
use moon_workspace::WorkspaceError;
use starbase_utils::fs::FsError;
use thiserror::Error;

#[derive(Error, Debug, Diagnostic)]
pub enum PipelineError {
    #[error("{0}")]
    Aborted(String),

    #[error("An unknown action was encountered in the pipeline. Unable to proceed!")]
    UnknownActionNode,

    #[diagnostic(transparent)]
    #[error(transparent)]
    DepGraph(#[from] DepGraphError),

    #[diagnostic(transparent)]
    #[error(transparent)]
    Fs(#[from] FsError),

    #[diagnostic(transparent)]
    #[error(transparent)]
    Moon(#[from] MoonError),

    #[diagnostic(transparent)]
    #[error(transparent)]
    Project(#[from] ProjectError),

    #[diagnostic(transparent)]
    #[error(transparent)]
    Runner(#[from] RunnerError),

    #[diagnostic(transparent)]
    #[error(transparent)]
    Target(#[from] TargetError),

    #[diagnostic(transparent)]
    #[error(transparent)]
    Tool(#[from] ToolError),

    #[diagnostic(transparent)]
    #[error(transparent)]
    Workspace(#[from] WorkspaceError),
}
