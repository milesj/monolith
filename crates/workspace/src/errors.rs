use moon_config::{constants, ValidationErrors};
use moon_error::MoonError;
use moon_project::ProjectError;
use moon_toolchain::ToolchainError;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum WorkspaceError {
    #[error("A dependency cycle has been detected for <file>{0}</file>.")]
    DepGraphCycleDetected(String),

    #[error("Unknown node {0} found in dependency graph. How did this get here?")]
    DepGraphUnknownNode(usize),

    #[error("{0}")]
    ActionRunnerFailure(String),

    #[error(
        "Unable to determine workspace root. Please create a <file>{}</file> configuration folder.",
        constants::CONFIG_DIRNAME
    )]
    MissingConfigDir,

    #[error(
        "Unable to locate a root <file>package.json</file>. Please create one alongside the <file>{}</file> configuration folder.",
        constants::CONFIG_DIRNAME
    )]
    MissingPackageJson,

    #[error(
        "Unable to locate <file>{}/{}</file> configuration file.",
        constants::CONFIG_DIRNAME,
        constants::CONFIG_WORKSPACE_FILENAME
    )]
    MissingWorkspaceConfigFile,

    #[error(
        "Failed to validate <file>{}/{}</file> configuration file.\n\n<muted>{0}</muted>",
        constants::CONFIG_DIRNAME,
        constants::CONFIG_WORKSPACE_FILENAME
    )]
    InvalidWorkspaceConfigFile(ValidationErrors),

    #[error(
        "Failed to validate <file>{}/{}</file> configuration file.\n\n<muted>{0}</muted>",
        constants::CONFIG_DIRNAME,
        constants::CONFIG_PROJECT_FILENAME
    )]
    InvalidGlobalProjectConfigFile(ValidationErrors),

    #[error(transparent)]
    Moon(#[from] MoonError),

    #[error(transparent)]
    Project(#[from] ProjectError),

    #[error(transparent)]
    Toolchain(#[from] ToolchainError),
}
