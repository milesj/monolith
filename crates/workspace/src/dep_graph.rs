use crate::errors::WorkspaceError;
use moon_logger::{color, debug, trace};
use moon_project::{ProjectGraph, Target, TouchedFilePaths};
use petgraph::algo::toposort;
use petgraph::dot::{Config, Dot};
use petgraph::graph::DiGraph;
use petgraph::Graph;
use std::collections::HashMap;

pub use petgraph::graph::NodeIndex;

pub enum NodeType {
    InstallNodeDeps,
    RunTarget(String), // target id
    SetupToolchain,
    SyncProject(String), // project id
}

pub struct Node {
    pub label: String,
    pub type_of: NodeType,
}

type GraphType = DiGraph<Node, ()>;

/// A directed acyclic graph (DAG) for the work that needs to be processed, based on a
/// project or task's dependency chain. This is also known as a "task graph" (not to
/// be confused with ours) or a "dependency graph". We call it a "work graph", as it's
/// the combination of those 2 with additional information for the work unit (a job).
pub struct DepGraph {
    graph: GraphType,

    /// Reference node for the "setup toolchain" job.
    toolchain_node: NodeIndex,

    /// Mapping of IDs to existing node indices.
    index_cache: HashMap<String, NodeIndex>,

    /// Reference node for the "install deps" job.
    install_deps_node: NodeIndex,
}

impl DepGraph {
    pub fn default() -> Self {
        debug!(
            target: "moon:dep-graph",
            "Creating work graph",
        );

        let mut graph: GraphType = Graph::new();

        // Toolchain must be setup first
        let toolchain_node = graph.add_node(Node {
            label: String::from("SetupToolchain"),
            type_of: NodeType::SetupToolchain,
        });

        // Deps can be installed *after* the toolchain exists
        let install_deps_node = graph.add_node(Node {
            label: String::from("InstallNodeDeps"),
            type_of: NodeType::InstallNodeDeps,
        });

        graph.add_edge(install_deps_node, toolchain_node, ());

        DepGraph {
            graph,
            index_cache: HashMap::new(),
            install_deps_node,
            toolchain_node,
        }
    }

    pub fn get_node_from_index(&self, index: NodeIndex) -> Option<&Node> {
        self.graph.node_weight(index)
    }

    pub fn sort_topological(&self) -> Result<Vec<NodeIndex>, WorkspaceError> {
        match toposort(&self.graph, None) {
            Ok(nodes) => Ok(nodes),
            Err(error) => Err(WorkspaceError::CycleDetected(error.node_id().index())),
        }
    }

    pub fn run_target(
        &mut self,
        target: &str,
        projects: &ProjectGraph,
    ) -> Result<NodeIndex, WorkspaceError> {
        if self.index_cache.contains_key(target) {
            return Ok(*self.index_cache.get(target).unwrap());
        }

        trace!(
            target: "moon:dep-graph",
            "Target {} does not exist in the work graph, inserting",
            color::id(target),
        );

        let (project_id, task_id) = Target::parse(target)?;
        let project = projects.get(&project_id)?;

        let node = self.graph.add_node(Node {
            label: format!("RunTarget({})", target),
            type_of: NodeType::RunTarget(target.to_owned()),
        });
        self.graph.add_edge(node, self.install_deps_node, ());

        // We should sync projects *before* running targets
        let project_node = self.sync_project(&project.id, projects)?;
        self.graph.add_edge(node, project_node, ());

        // And we also need to wait on all dependent nodes
        let task = project.get_task(&task_id)?;

        if !task.deps.is_empty() {
            let dep_names: Vec<String> = task
                .deps
                .clone()
                .into_iter()
                .map(|d| color::symbol(&d))
                .collect();

            trace!(
                target: "moon:dep-graph",
                "Adding dependencies {} from target {}",
                dep_names.join(", "),
                color::id(target),
            );

            for dep_target in &task.deps {
                let dep_node = self.run_target(dep_target, projects)?;
                self.graph.add_edge(node, dep_node, ());
            }
        }

        // Also cache so we don't run the same target multiple times
        self.index_cache.insert(target.to_owned(), node);

        Ok(node)
    }

    pub fn run_target_if_touched(
        &mut self,
        target: &str,
        touched_files: &TouchedFilePaths,
        projects: &ProjectGraph,
    ) -> Result<Option<NodeIndex>, WorkspaceError> {
        // Validate project first
        let (project_id, task_id) = Target::parse(target)?;
        let project = projects.get(&project_id)?;

        if !project.is_affected(touched_files) {
            trace!(
                target: "moon:dep-graph",
                "Project {} not affected based on touched files, skipping",
                color::id(&project_id),
            );

            return Ok(None);
        }

        // Validate task exists for project
        let task = project.get_task(&task_id)?;

        if !task.is_affected(touched_files)? {
            trace!(
                target: "moon:dep-graph",
                "Project {} task {} not affected based on touched files, skipping",
                color::id(&project_id),
                color::id(&task_id),
            );

            return Ok(None);
        }

        Ok(Some(self.run_target(target, projects)?))
    }

    pub fn sync_project(
        &mut self,
        project_id: &str,
        projects: &ProjectGraph,
    ) -> Result<NodeIndex, WorkspaceError> {
        if self.index_cache.contains_key(project_id) {
            return Ok(*self.index_cache.get(project_id).unwrap());
        }

        trace!(
            target: "moon:dep-graph",
            "Syncing project {} configs and dependencies",
            color::id(project_id),
        );

        let project = projects.get(project_id)?;

        // Sync can be run in parallel while deps are installing
        let node_index = self.graph.add_node(Node {
            label: format!("SyncProject({})", project_id),
            type_of: NodeType::SyncProject(project_id.to_owned()),
        });

        self.graph.add_edge(node_index, self.toolchain_node, ());

        // But we need to wait on all dependent nodes
        for dep_id in projects.get_dependencies_of(&project)? {
            let dep_node_index = self.sync_project(&dep_id, projects)?;
            self.graph.add_edge(node_index, dep_node_index, ());
        }

        // Also cache so we don't sync the same project multiple times
        self.index_cache.insert(project_id.to_owned(), node_index);

        Ok(node_index)
    }

    pub fn to_dot(&self) -> String {
        let graph = self.graph.map(|_, n| n.label.as_str(), |_, e| e);
        let dot = Dot::with_config(&graph, &[Config::EdgeNoLabel]);

        format!("{:?}", dot)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use moon_config::GlobalProjectConfig;
    use moon_project::ProjectGraph;
    use moon_utils::test::get_fixtures_dir;
    use std::collections::HashMap;

    fn create_project_graph() -> ProjectGraph {
        ProjectGraph::new(
            &get_fixtures_dir("projects"),
            GlobalProjectConfig::default(),
            &HashMap::from([
                ("advanced".to_owned(), "advanced".to_owned()),
                ("basic".to_owned(), "basic".to_owned()),
                ("emptyConfig".to_owned(), "empty-config".to_owned()),
                ("noConfig".to_owned(), "no-config".to_owned()),
                ("foo".to_owned(), "deps/foo".to_owned()),
                ("bar".to_owned(), "deps/bar".to_owned()),
                ("baz".to_owned(), "deps/baz".to_owned()),
                ("tasks".to_owned(), "tasks".to_owned()),
            ]),
        )
    }

    #[test]
    fn default_graph() {
        let graph = DepGraph::default();

        assert_snapshot!(graph.to_dot());
    }

    mod run_target {
        use super::*;

        #[test]
        fn single_targets() {
            let projects = create_project_graph();

            let mut graph = DepGraph::default();
            graph.sync_project("advanced", &projects).unwrap();
            graph.sync_project("basic", &projects).unwrap();
            graph.sync_project("emptyConfig", &projects).unwrap();
            graph.sync_project("noConfig", &projects).unwrap();

            assert_snapshot!(graph.to_dot());
        }
    }

    mod sync_project {
        use super::*;

        #[test]
        fn isolated_projects() {
            let projects = create_project_graph();

            let mut graph = DepGraph::default();
            graph.sync_project("advanced", &projects).unwrap();
            graph.sync_project("basic", &projects).unwrap();
            graph.sync_project("emptyConfig", &projects).unwrap();
            graph.sync_project("noConfig", &projects).unwrap();

            assert_snapshot!(graph.to_dot());
        }

        #[test]
        fn projects_with_deps() {
            let projects = create_project_graph();

            let mut graph = DepGraph::default();
            graph.sync_project("foo", &projects).unwrap();
            graph.sync_project("bar", &projects).unwrap();
            graph.sync_project("baz", &projects).unwrap();
            graph.sync_project("basic", &projects).unwrap();

            assert_snapshot!(graph.to_dot());
        }

        #[test]
        fn projects_with_tasks() {
            let projects = create_project_graph();

            let mut graph = DepGraph::default();
            graph.sync_project("noConfig", &projects).unwrap();
            graph.sync_project("tasks", &projects).unwrap();

            assert_snapshot!(graph.to_dot());
        }
    }
}
