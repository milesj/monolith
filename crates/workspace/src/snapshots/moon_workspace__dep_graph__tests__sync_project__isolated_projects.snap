---
source: crates/workspace/src/dep_graph.rs
assertion_line: 288
expression: graph.to_dot()

---
digraph {
    0 [ label = "\"SetupToolchain\"" ]
    1 [ label = "\"InstallNodeDeps\"" ]
    2 [ label = "\"SyncProject(advanced)\"" ]
    3 [ label = "\"SyncProject(basic)\"" ]
    4 [ label = "\"SyncProject(noConfig)\"" ]
    5 [ label = "\"SyncProject(emptyConfig)\"" ]
    1 -> 0 [ ]
    2 -> 0 [ ]
    3 -> 0 [ ]
    4 -> 0 [ ]
    3 -> 4 [ ]
    5 -> 0 [ ]
}

